### PR TITLE
Integrate Glimmer Macros For Binary VM

### PR DESCRIPTION
**Pull request by chadhietala**
This moves Ember over to the new Binary VM in glimmer (Requires [this branch](https://github.com/tildeio/glimmer/commits/runtime-null-checks)). Notable changes:

- `{{debugger}}` is now implemented as a built-in Glimmer statement syntax due to the fact it needs access to the `SymbolTable` and other information from the VM. Also we now hard error if the debugger is passed an expression, from what I can tell this just failed silently is we did nothing with the positional or named params.
- `refineStatement` on the `Environment` is no more in favor of `populateMacros` and `block|inlines.addMissing`.
- No more syntax objects but now byte arrays. 😎

### CODE REVIEW

File: packages/ember-glimmer/lib/syntax/dynamic-component.js [hunk 1]
```diff
@@ -16,37 +12,33 @@ function dynamicComponentFor(vm, symbolTable) {
...
+  let definitionArgs = [params.slice(0, 1), null, null, null];
+  let args = [params.slice(1), hash, _default, inverse];
+  builder.component.dynamic(definitionArgs, dynamicComponentFor, args, builder.symbolTable);
+  return true;
+}
 
-  compile(builder) {
-    builder.component.dynamic(this.definitionArgs, this.definition, this.args, this.symbolTable, this.shadow);
-  }
+export function inlineComponentMacro(path, params, hash, builder) {
```
Comment (chadhietala) [inline packages/ember-glimmer/lib/syntax/dynamic-component.js]: This likely can go away in favor of `dynamicComponentMacro`.

File: ember-cli-build.js [hunk 2]
```diff
@@ -197,26 +221,26 @@ function babelConfigFor(environment) {
...
-  }), {
-    files: ['**/*.js'],
-    pattern: {
-      match: /\/\/#\s+sourceMappingURL.*/g,
-      replacement: ''
-    },
-    annotation: 'strip sourceMappingURL'
-  });
-}
+// var glimmerEngine = require('glimmer-engine/ember-cli-build')({
```
Comment (rwjblue) [inline ember-cli-build.js]: We can probably just delete it right?

File: packages/ember-glimmer/lib/helpers/component.js [hunk 3]
```diff
@@ -206,6 +205,11 @@ function createCurriedDefinition(definition, args) {
   );
 }
 
+const EMPTY_BLOCKS = {
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/helpers/component.js]: Should this be frozen?

File: packages/ember-glimmer/lib/syntax.js [hunk 4]
```diff
@@ -1,44 +1,85 @@
...
+  if (path.length > 1) {
+    return closureComponentMacro(path, params, hash, null, null, builder);
+  }
 
-export function findSyntaxBuilder(key) {
-  let index = syntaxKeys.indexOf(key);
+  let { symbolTable } = builder;
+
+  let definition;
+  if (name.indexOf('-') > -1) {
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/syntax.js]: Should we use the contains dash caching here?

File: packages/ember-glimmer/lib/syntax/-in-element.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { BaselineSyntax, compileArgs } from '@glimmer/runtime';
+import { unwrap } from '@glimmer/util';
+
+const {
+  defaultBlock,
+  params,
+  hash
+} = BaselineSyntax.NestedBlock;
+
+export function _inElementMacro(sexp, builder) {
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/syntax/-in-element.js]: Why is this moved into Ember itself? I thought it was baked into Glimmer?

File: packages/ember-glimmer/lib/syntax.js [hunk 6]
```diff
@@ -1,44 +1,85 @@
...
+import { assert } from 'ember-metal';
 
+function refineInlineSyntax(path, params, hash, builder) {
+  let [ name ] = path;
 
-let syntaxKeys = [];
-let syntaxes = [];
+  assert(`You attempted to overwrite the built-in helper "${name}" which is not allowed. Please rename the helper.`, !(builder.env.builtInHelpers[name] && builder.env.owner.hasRegistration(`helper:${name}`)));
 
-export function registerSyntax(key, syntax) {
```
Comment (rwjblue) [inline packages/ember-glimmer/lib/syntax.js]: I would still like to export a mechanism to register macros. Without one we loose the ability to experiment with different syntaxes (e.g. `ember-let`). 

It seems totally good to change it to a more generic thing like `registerMacro`, but we shouldn't remove the ability to customize/experiment.

File: package.json [hunk 7]
```diff
@@ -27,6 +27,11 @@
     "test:testem": "testem -f testem.dist.json"
   },
   "dependencies": {
+    "@glimmer/compiler": "^0.21.1",
+    "@glimmer/node": "^0.21.1",
+    "@glimmer/reference": "^0.21.0",
```
Comment (chadhietala) [inline package.json]: No I don't believe so. The way `lerna publish` works is that it looks if packages were effected in between the `HEAD` and the last tag if its not effected it will not publish a new version.

File: packages/ember-glimmer/lib/syntax/-in-element.js (hunk 5 above)
Comment (chadhietala) [inline packages/ember-glimmer/lib/syntax/-in-element.js]: Seems like it was removed from Glimmer.

File: packages/ember-glimmer/lib/syntax.js (hunk 6 above)
Comment (chadhietala) [inline packages/ember-glimmer/lib/syntax.js]: I added an API for this. Let me know what you think.

File: package.json [hunk 8]
```diff
@@ -35,12 +40,14 @@
     "ember-cli-test-info": "^1.0.0",
     "ember-cli-valid-component-name": "^1.0.0",
     "ember-cli-version-checker": "^1.1.7",
+    "handlebars": "~3",
```
Comment (chadhietala) [inline package.json]: @krisselden might want to chime in here. I don't remember why we did this.

File: packages/ember-glimmer/lib/helpers/component.js (hunk 3 above)
Comment (krisselden) [inline packages/ember-glimmer/lib/helpers/component.js]: at least in dev mode, freezing it does make it a different "shape" polymorphic with the unfrozen versions, not sure if it is already polymorphic and empty and non empty should be both seen by the time it is JITed so may not matter.

**Comment by rwjblue** [comment]
Awesome job here @chadhietala!
